ckanext/datastore/helpers.py

```diff
@@ -26,10 +26,12 @@ def is_valid_field_name(name):
     * can't start with underscore
     * can't contain double quote (")
     * can't be empty
+    * can't be longer than 63 characters
     '''
     return (name and name == name.strip() and
             not name.startswith('_') and
-            '"' not in name)
+            '"' not in name and
+            not len(name) > 63)
```

Comment:
Does postgres count character length the same way we do in python?

Does the encoding of the datastore database affect this calculation?

Reply:
I believe so.

In my testing, I encountered this problem when I was automatically creating aliases when a resource was created using the template "ownerorg-packagename-resourcename"(https://github.com/ckan/ideas/issues/264), and sometimes, it will create aliases longer than 63 characters and the resulting aliases are truncated at 63 characters by postgres.

Also, there is another area where the datastore checks for 63 characters already - 
https://github.com/ckan/ckan/blob/cc000d3acf7401957d8321be53e8cabc2d9ebf3a/ckanext/datastore/backend/postgres.py#L863-L867

However, it only does it for field names. Not table or alias names.

Reply:
Both might be wrong. It sounds like the identifier length is limited to 63 bytes, so we either need to know the encoding of the DB  or we should do a  `select octet_length(...)` on the strings we plan to use field or alias names.

Reply:
Since postgres UTF8 encoding is already required, and the code has the `# encoding utf-8` magic comment, won't doing `len()` be good enough?

If not, I'd go for `len(name.encode("utf-8"))` instead of querying postgres for `octet_length(...)`.

Reply:
where is postgres utf-8 encoding required? There's nothing preventing a user from setting the datastore encoding to cp1252 to increase performance or space-efficiency for their data is there?

Reply:
The docs **_imply_** a utf-8 requirement:

https://docs.ckan.org/en/latest/maintaining/installing/install-from-source.html?highlight=UTF#setup-a-postgresql-database
https://docs.ckan.org/en/latest/maintaining/datastore.html?highlight=UTF#set-up-the-database

There is also a hardcoded utf8 assumption in the postgres backend.
https://github.com/ckan/ckan/blob/cc000d3acf7401957d8321be53e8cabc2d9ebf3a/ckanext/datastore/backend/postgres.py#L555-L557

Regardless, the docs should be updated to state if utf8 encoding is actually not required, and maybe even the tests updated for alternate encodings.

Reply:
This `encode` is used to generate a hexadecimal hash, it doesn't affect any postgres encoding.

In other issues we're talking about letting users create their own materialized views or foreign data wrappers or other objects. It's not unreasonable for a user to want an encoding optimal for their use case.

```sql
postgres=# create database cp1252test encoding 'win1252' lc_ctype='C' lc_collate='C' template=template0;
CREATE DATABASE
cp1252test=# create table t1 ( ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ text );
NOTICE:  identifier "ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ" will be truncated to "ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþ"
```

vs

```
utf8test=# create table t1 ( ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ text );
NOTICE:  identifier "ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ" will be truncated to "ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞ"
CREATE TABLE
```

Reply:
Point taken!  Especially as someone who's asking for more advanced postgres support. :)

And speaking of advanced postgres support, perhaps, instead of hardcoding it, get it from postgres directly with `pg_control_init()` now that postgres v9.5 is about to [get EOLed Feb 11, 2021](https://endoflife.software/applications/databases/postgresql).

https://pgpedia.info/p/pg_control_init.html